src/dtx/dtx_common.c

```diff
@@ -1111,45 +1112,66 @@ dtx_leader_begin(daos_handle_t coh, struct dtx_id *dti,
 		 uint32_t pm_ver, daos_unit_oid_t *leader_oid,
 		 struct dtx_id *dti_cos, int dti_cos_cnt,
 		 struct daos_shard_tgt *tgts, int tgt_cnt, uint32_t flags,
+		 struct obj_coll_target *coll_tgts, int coll_tgt_nr,
 		 struct dtx_memberships *mbs, struct dtx_leader_handle **p_dlh)
 {
 	struct dtx_leader_handle	*dlh;
 	struct dtx_handle		*dth;
+	int				 cnt;
 	int				 rc;
 	int				 i;
 
-	D_ALLOC(dlh, sizeof(*dlh) + sizeof(struct dtx_sub_status) * tgt_cnt);
+	if (coll_tgt_nr > 0)
+		/* coll_tgts[0] is for the target(s) on the leader rank. */
+		cnt = coll_tgt_nr - 1;
+	else
+		cnt = tgt_cnt;
+
+	D_ALLOC(dlh, sizeof(*dlh) + sizeof(struct dtx_sub_status) * cnt);
 	if (dlh == NULL)
 		return -DER_NOMEM;
 
-	if (tgt_cnt > 0) {
+	if (cnt > 0) {
 		dlh->dlh_future = ABT_FUTURE_NULL;
 		dlh->dlh_subs = (struct dtx_sub_status *)(dlh + 1);
-		for (i = 0; i < tgt_cnt; i++) {
-			dlh->dlh_subs[i].dss_tgt = tgts[i];
-			if (unlikely(tgts[i].st_flags & DTF_DELAY_FORWARD))
-				dlh->dlh_delay_sub_cnt++;
+
+		if (coll_tgt_nr > 0) {
+			for (i = 0; i < cnt; i++)
+				dlh->dlh_subs[i].dss_tgt.st_rank = coll_tgts[i + 1].oct_rank;
+
+			/* NOTE: currently do not support DTF_DELAY_FORWARD for collective DTX. */
```

Comment:
just confirm, what it means if do obj_punch() with DAOS_COND_PUNCH flag? DAOS_COND_PUNCH only meaningful for DKEY/AKEY punch?

Reply:
Yes, the conditional punch is only applicable for dkey/akey, not on object.